src/components/Frame/components/Loading/Loading.tsx

```diff
@@ -1,78 +1,69 @@
-import React from 'react';
+import React, {useEffect, useState} from 'react';
 import debounce from 'lodash/debounce';
+import {useLazyRef} from '../../../../utilities/use-lazy-ref';
 import styles from './Loading.scss';
 
-export interface LoadingProps {}
-
-interface State {
-  progress: number;
-  step: number;
-  animation: number | null;
-}
-
 const INITIAL_STEP = 10;
 const STUCK_THRESHOLD = 99;
 
-export class Loading extends React.Component<LoadingProps, State> {
-  state: State = {
-    progress: 0,
-    step: INITIAL_STEP,
-    animation: null,
-  };
-
-  private ariaValuenow = debounce(() => {
-    const {progress} = this.state;
-    return Math.floor(progress / 10) * 10;
-  }, 15);
+export interface LoadingProps {}
 
-  componentDidMount() {
-    this.increment();
-  }
+export function Loading(_props: LoadingProps) {
+  const [progress, setProgress] = useState(0);
 
-  componentWillUnmount() {
-    const {animation} = this.state;
+  useEffect(() => {
+    let animation: number | undefined;
+    let step = INITIAL_STEP;
+    let currentProgress = 0;
+    let previusProgress = 0;
 
-    if (animation != null) {
-      cancelAnimationFrame(animation);
-    }
-  }
+    const increment = () => {
+      if (currentProgress >= STUCK_THRESHOLD) {
+        return;
+      }
 
-  render() {
-    const {progress} = this.state;
+      currentProgress = Math.min(currentProgress + step, 100);
+      const nextProgress = Math.floor(currentProgress) / 100;
 
-    const customStyles = {
-      transform: `scaleX(${Math.floor(progress) / 100})`,
-    };
+      if (nextProgress !== previusProgress) {
+        setProgress(nextProgress);
+        previusProgress = nextProgress;
+      }
 
-    const ariaValuenow = this.ariaValuenow();
+      step = INITIAL_STEP ** -(currentProgress / 25);
 
-    return (
-      <div className={styles.Loading}>
-        <div
-          className={styles.Level}
-          style={customStyles}
-          aria-valuenow={ariaValuenow}
-          aria-valuemin={0}
-          aria-valuemax={100}
-          role="progressbar"
-        />
-      </div>
-    );
-  }
+      animation = requestAnimationFrame(increment);
+    };
 
-  private increment() {
-    const {progress, step} = this.state;
+    increment();
 
-    if (progress >= STUCK_THRESHOLD) {
-      return;
-    }
+    return () => {
+      if (animation) {
+        cancelAnimationFrame(animation);
+      }
+    };
+  }, []);
 
-    const animation = requestAnimationFrame(() => this.increment());
+  const customStyles = {
+    transform: `scaleX(${progress})`,
+  };
 
-    this.setState({
-      progress: Math.min(progress + step, 100),
-      step: INITIAL_STEP ** -(progress / 25),
-      animation,
-    });
-  }
+  const ariaValuenow = useLazyRef(() =>
+    debounce((progress: number) => {
```

Comment:
In this aproach component will only re-rendered when progress changes, unlike the old aproach which would re-render with each `requestAnimationFrame` call. I guess it'd be safe to remove debounce and calculate the valenow inline.

Reply:
This is a a pure function  that doesn't depend on anything in the outer closure, so there's no need to put it inside the component or do any of the lazyRef stuff, just define it as a regular function outside of the component function

---

That said, now that we're only rendering this once per progress change I'm good with just defining this value inline. If I've understood your other changes - where progress is now a number between 0 and 1, rounded to two decimal places this should do:

```
aria-valuenow={progress * 100}
```

Reply:
I'll refactor it and remove `debounce`, just FTR if we move that function outer closure, multiple instance of this component will use same debounced ref and it might cause data leak between those instance (they mill use each other values).